### PR TITLE
Drivetrain memory fixes

### PR DESCRIPTION
**Pull request by awfulbananas**
greatly reduces the number of vectors initialized per cycle

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: I recommend instead making this a static member of the Vector class. We call this the zero vector. May I recommend `Vector.Zero`?

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java:316]: I believe you shouldn't modify the buffer given to you. You should make a private buffer in the class called `moveSpd`

**Review by EmDash00 — CHANGES_REQUESTED** [review]
You should stop allocating the measure Angle as well. See the comment I left you on slack about using a `MutableMeasure`.

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Should be private if it's going to live here.

File: src/main/java/frc/robot/RobotContainer.java
Comment (garrettwitowski) [inline src/main/java/frc/robot/RobotContainer.java]: Why is this here?

**Review by garrettwitowski — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java
Comment (awfulbananas) [reply src/main/java/frc/robot/RobotContainer.java]: it was there after resolving rebase conflicts, if it should be removed it should be from drivetrain rather than this branch

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: I think I'll go with emmy's suggestion to put it in the Vector class

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: that's probably a good idea so I'll do it, though my only worry is that then any code from anywhere can effectively just modify the location of zero

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:316]: what do you mean by a buffer?

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: although, I do find that kind of interesting as a concept, so maybe it's fine

Comment (garrettwitowski) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Vector.Zero should be marked final. Not sure if you can return const/final equivalent from a method in Java?

Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: I can, but Vector is mutable

Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: or are you asking if I have an immutable version of Vector, in which case I could, but currently don't

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:316]: A buffer is a region of memory that is used to store some things while it is being moved around. A `Vector` is in this necessarily is a buffer as that's what it does. I don't think your drive algorithm should modify the input given to it. You should instead allocate another vector to copy the input into so that you can mutate that instead. This kind of modification can lead to bizarre bugs if you make assumptions about the state of the input vector after the call...

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Hm, @garrettwitowski he means that it'd be like marking an array final. Marking an array final doesn't prevent you from modifying its contents. Yes I think there should be an ImmutableVector, but that is for a future time. For now we should just avoid modifying it...

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Why not just `return m_heading.mut_replace(m_gyro.getAngle(), Units.Degrees);`?

**Review by EmDash00 — CHANGES_REQUESTED** [review]

**Comment by EmDash00** [comment]
Not sure if you noticed, but the commit that moved the origin into the `Vector` class also fixed the rotations. Fixing the rotations should have been a commit on the drivetrain branch itself...

```
git reset HEAD^
# only commit the hunk that contains the zero vector
git push --force
```

Please have the Code Integration lead review before the push and to backup local copies.

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:316]: I know, but I checked carefully, and the whole point of this branch is to save memory by not allocating as many vectors

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: I don't actually know, though this way it is obvious that the value of the measure is being changed

**Comment by awfulbananas** [comment]
> Not sure if you noticed, but the commit that moved the origin into the `Vector` class also fixed the rotations. Fixing the rotations should have been a commit on the drivetrain branch itself...
> 
> ```
> git reset HEAD^
> # only commit the hunk that contains the zero vector
> git push --force
> ```
> 
> Please have the Code Integration lead review before the push and to backup local copies.

I'm not sure how it happened, but the .rot() fix is already it's own commit in drivetrain, and github doesn't detect any merge conflicts so I think it's probably fine

File: src/main/java/frc/utils/Vector.java
Comment (garrettwitowski) [inline src/main/java/frc/utils/Vector.java]: nit: prefer the explicit constructor here so we don't have to depend on the implementation of the default constructor (which could change at any time).

**Comment by garrettwitowski** [comment]
Hmm why are we merging this to drivetrain rather than develop?

**Comment by garrettwitowski** [comment]
> > Not sure if you noticed, but the commit that moved the origin into the `Vector` class also fixed the rotations. Fixing the rotations should have been a commit on the drivetrain branch itself...
> > ```
> > git reset HEAD^
> > # only commit the hunk that contains the zero vector
> > git push --force
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please have the Code Integration lead review before the push and to backup local copies.
> 
> I'm not sure how it happened, but the .rot() fix is already it's own commit in drivetrain, and github doesn't detect any merge conflicts so I think it's probably fine

It looks like your most recent commit (to use MutableMeasure) introduced this .rot() change here. I'm surprised that there are no merge conflicts here.

File: src/main/java/frc/utils/Vector.java
Comment (awfulbananas) [reply src/main/java/frc/utils/Vector.java]: since I made the default constructor, I know it will always be (0, 0), but I can make it more explicit

**Comment by awfulbananas** [comment]
> Hmm why are we merging this to drivetrain rather than develop?

the changes are specific to drivetrain, since Vectors were mainly being used there, and they were a relatively large memory sink (allocating ~6 Vectors per cycle the drivetrain was active), it could go into develop, but it was based on drivetrain, so it should probably be merged into drivetrain

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:316]: > I know, but I checked carefully, and the whole point of this branch is to save memory by not allocating as many vectors

Allocating one more vector at construction isn't a high cost. I think you should do it as to not mutate the input vector.

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (EmDash00) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: > I don't actually know, though this way it is obvious that the value of the measure is being changed

I think it's already evident since you're using `mut_replace` I don't like the self-overwriting. Can you change that?

Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: I'll fix it

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:316]: sounds good

File: src/main/java/frc/utils/Vector.java
Comment (EmDash00) [reply src/main/java/frc/utils/Vector.java]: I think it'd be better to use 0, 0 to make it more futureproof since we plan on reusing this code. A lot of code in the real world is written this way to be resistant to changes in dependencies.

**Comment by awfulbananas** [comment]
I've resolved all the requested changes, could I get one final look over before merging?

File: src/main/java/frc/robot/RobotContainer.java
Comment (ChainedGCH) [inline src/main/java/frc/robot/RobotContainer.java]: Why is your X set to Y?

Comment (awfulbananas) [reply src/main/java/frc/robot/RobotContainer.java]: That’s a bug, I’ll fix it

Comment (garrettwitowski) [reply src/main/java/frc/robot/RobotContainer.java]: Are you sure that's a bug?

File: src/main/java/frc/robot/RobotContainer.java
Comment (garrettwitowski) [reply src/main/java/frc/robot/RobotContainer.java]: Please remove it from from here. It doesn't belong here or in drivetrain.

File: src/main/java/frc/robot/RobotContainer.java
Comment (garrettwitowski) [inline src/main/java/frc/robot/RobotContainer.java]: Remove please

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java:152]: This might have the unintended side-effect of rotating to 0 degrees.

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java:328]: Why is this removed?

**Review by garrettwitowski — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:152]: The move method takes a direction to move in and an amount to rotate, not a target angle

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java
Comment (awfulbananas) [reply src/main/java/frc/robot/subsystems/drive/Drivetrain.java:328]: I moved it to above the slew rate limiting to prevent a future bug where the slew rate limiting would apply to the proportional input (a range of -1 to 1) instead of the actual movement speed range (-max speed to max speed)

File: src/main/java/frc/robot/RobotContainer.java
Comment (awfulbananas) [reply src/main/java/frc/robot/RobotContainer.java]: I just checked the old code and I guess it's not, though it might be cleaner to rotate the vector rather than swapping the x and y inputs

File: src/main/java/frc/robot/RobotContainer.java
Comment (awfulbananas) [reply src/main/java/frc/robot/RobotContainer.java]: fixed

File: src/main/java/frc/robot/RobotContainer.java
Comment (awfulbananas) [reply src/main/java/frc/robot/RobotContainer.java]: done

**Comment by awfulbananas** [comment]
removing vision from RobotContainer caused a merge conflict, I'll resolve that then wait for some reviews before merging

File: src/main/java/frc/robot/RobotContainer.java
Comment (garrettwitowski) [reply src/main/java/frc/robot/RobotContainer.java]: I agree that it can (and should) be more clear that this is a transformation from "Joystick XY" to "Robot XY", which is really a projection into the standard 3D coordinate system.

Comment (garrettwitowski) [reply src/main/java/frc/robot/RobotContainer.java]: @ChainedGCH this is because we're transforming "Joystick XY" into "Robot XY(Z)". Robots operate in 3-D space and so use that coordinate system.

Comment (EmDash00) [reply src/main/java/frc/robot/RobotContainer.java]: @ChainedGCH Yes their "x" is a different X. It's the same X as in the right hand rule. There are many different choices one can make, but so long as it's a rotated version of your "right-hand rule" fingers, it's positive orientation. Other orientations are mirrored versions of your right hand rule and are "left-handed" and thus negative orientation. You can in general use either, but most algorithms and formulas use positive orientation, so it tends to be more convenient. Long story short, the robot coordinate system has X as back-forward, Y as left-right, and Z as up-down. This is a common coordinate system. 

@awfulbananas I think this code should've had a comment mentioning this to avoid this confusion. There's nothing wrong with not rotating explicitly as that's more computationally expensive.

**Review by ChainedGCH — COMMENTED** [review]
Have you resolved the merge conflict(s)? There's a list of instructions in slack by Garrett to reference

**Review by ChainedGCH — APPROVED** [review]
Problems have been fixed (to my knowledge)

**Review by Irishumanoid — APPROVED** [review]

**Pull request merged**
